### PR TITLE
DataArtifact add size/type method

### PR DESCRIPTION
**Pull request by SteNicholas**
DataArtifact add size/type method to tell the size of the data contained in this artifact and type for this artifact.

### CODE REVIEW

**Comment by savingoyal** [comment]
@SteNicholas Thanks for the PR! Much appreciated! Please refer to the [contributing guide](https://github.com/Netflix/metaflow/blob/master/CONTRIBUTING.md). We are aligning internally on how to make contributing to Metaflow easy and transparent.

File: metaflow/client/core.py [hunk 1]
```diff
@@ -680,13 +685,33 @@ def data(self):
...
+            Object size contained in this artifact
+        """
+        ds_type = self._object['ds_type']
+        sha = self._object['sha']
+        with filecache.get_data(ds_type, self.path_components[0], sha) as f:
+            obj = pickle.load(f)
+            return sys.getsizeof(obj)
+
+    @property
+    def type(self):
```
Comment (romain-intel) [inline metaflow/client/core.py]: This function looks fine.

File: metaflow/client/core.py [hunk 2]
```diff
@@ -680,13 +685,33 @@ def data(self):
...
-    # @property
-    # def type(self)
+        Returns
+        -------
+        size
+            Object size contained in this artifact
+        """
+        ds_type = self._object['ds_type']
+        sha = self._object['sha']
+        with filecache.get_data(ds_type, self.path_components[0], sha) as f:
```
Comment (romain-intel) [inline metaflow/client/core.py]: I am not a big fan of having to unpickle and then throw away the result just to get the size (and doing it every time). A potentially better approach would be to store the size information along with the sha, ds_type, etc and return that directly.

Comment (SteNicholas) [reply metaflow/client/core.py]: @romain-intel I couldn't agree with the approach you mentioned. The sha, ds_type, etc are stored in json. Storing the size information into json is unreasonable. This should calculate the size by `get_data`.

Comment (savingoyal) [reply metaflow/client/core.py]: @SteNicholas Why do you think storing the size information is unreasonable?

Comment (SteNicholas) [reply metaflow/client/core.py]: @savingoyal In my option, the sha, ds_type etc. information is stored in json file. And the size couldn't be defined the value in json.

Comment (savingoyal) [reply metaflow/client/core.py]: Why not?

Comment (SteNicholas) [reply metaflow/client/core.py]: @savingoyal Do you mean that the size should be defined the value in json? I will modify this by storing the size in json.

Comment (savingoyal) [reply metaflow/client/core.py]: yes, indeed.

Comment (SteNicholas) [reply metaflow/client/core.py]: @savingoyal I have already modified the size storing in info. Please review this again.

Comment (savingoyal) [reply metaflow/client/core.py]: Thanks, I will take a look in the morning.

**Comment by romain-intel** [comment]
Will look at this a little later today. Apologies for the delay.

**Comment by SteNicholas** [comment]
> Will look at this a little later today. Apologies for the delay.

When you have free time, please review this again. Thanks.

File: metaflow/datastore/datastore.py [hunk 3]
```diff
@@ -525,7 +525,8 @@ def _register_data_artifacts(self, objects, info):
                                   ds_type=self.TYPE,
                                   url=self.object_path(sha),
                                   sha=sha,
-                                  type=info[var]['encoding'])
+                                  type=info[var]['encoding'],
+                                  size=info[var]['size'])
```
Comment (romain-intel) [inline metaflow/datastore/datastore.py]: I believe the size that is in info corresponds to the size of the pickled object and I believe you want (which I agree with) to return the size of the unpickled object. You may have to modify _save_object in datastore.py to properly return the unpickled size.

Comment (SteNicholas) [reply metaflow/datastore/datastore.py]: @romain-intel Yeah, I made mistake in the return size of _save_object. And I have modified this unpickled object size. Please review this again.

File: metaflow/datastore/datastore.py [hunk 4]
```diff
@@ -441,9 +441,10 @@ def _save_object(self, transformable_obj, var, force_v4=False):
                                         % var)
                 transformable_obj.transform(lambda x: pickle.dumps(x, protocol=4))
         sha = sha1(transformable_obj.current()).hexdigest()
-        sz = len(transformable_obj.current())
-        self.save_data(sha, transformable_obj)
-        return sha, sz, blobtype
+        pickle_sz = len(transformable_obj.current())
+        path = self.save_data(sha, transformable_obj)
+        unpickle_sz = sys.getsizeof(open(path, 'rb'))
```
Comment (romain-intel) [inline metaflow/datastore/datastore.py]: Path is not necessarily local (could be on S3) so this won't work. Even if local, this may be a GZip path.

**Comment by SteNicholas** [comment]
> Apologies for the delay in re-reviewing this. My feeling right now is to punt on the size question (ie: not put anything related to size) and just keep the type notion. The reason for this is after a bit more thought, I wonder if we can ever return the unpickled size. There is no good way in Python to get the size of an object in Python (or at least no easy way). The pickled size may be a good approximation but I wonder if that has use to the user. The size on file (so pickled + gzip) may also have some value. In short, I think this may require a bit more thought to determine what "size" means for an artifact.

@romain-intel Therefore, in this interface, whether to provide user with `size` method? And about the size, I thought that the size should be defined that size is unpickled size of artifact object. Because in `DataArtifact`, the `data` method returns unpickled representation of the data contained in this artifact, so the `size` method should return unpickled size. Am I right?

**Comment by romain-intel** [comment]
size was integrated in another PR. This is probably also a bit old since the datastore changed a lot so a new one can be opened if needed.

**Pull request closed without merging**